Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+Add new library `nethost` which will act as the easy to use host for loading managed components.
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dynlib`). For ease of use there would be a header file for C++ apps as well as `.lib`/`.a` for easy linking.
+Apps using the component hosting functionality would ship this library as part ot the app. So similarly to `apphost`/`comhost`/`ijwhost` the `nethost` will be shipped as part of .NET Core SDK to be included in customer's projects.
+
+This library would expose two APIs
+
+### Load managed component and get a function pointer
+``` C++
+int nethost_load_assembly_method(
+        const char * assembly_path,
+        const char * type_name,
+        const char * method_name,
+        const void * reserved,
+        void ** delegate);
+)
+```
+This API will
+* Locate the assembly using the `assembly_path` and its `.runtimeconfig.json` and determine the frameworks it requires to run. (Note that only framework dependent components will be supported for now).
+* If the process doesn't have CoreCLR loaded (more specifically `hostpolicy` library)
+  * Using the `.runtimeconfig.json` resolve required frameworks (just like if running an application) and load the runtime.
+* Else the CoreCLR is already loaded, in that case validate that required frameworks for the component can be satisfied by the runtime.
+  * If the required frameworks are not already present, fail. No support to load additional frameworks for now.
+* Call into the runtime (`System.Private.CoreLib` specifically)
+  * Create a new isolated `AssemblyLoadContext` (possibly reusing for the same components)
+  * Load the component's assembly into it
+  * Find the requested `type_name` and `method_name`
+  * Return a native callable function pointer to the requested method
+
+The `reserved` argument is currently not used and must be set to `nullptr`.
+
+If the runtime is initialized by this function, it will only be populated with framework assemblies (its TPA), none of the component's assemblies will be loaded into the default context.
+
+*As proposed there would be no support for unloading components. For discussion on possible solutions see open issues below.*
+
+### Locate `hostfxr`
+``` C++
+typedef void(* nethost_get_hosfxr_path_result_fn)(const char * hostfxr_path);
+
+int nethost_get_hostfxr_path(
+        nethost_get_hostfxr_path_result_fn result,
+        const char * assembly_path); // Optional
+```
+
+This API locates the `hostfxr` and returns its path by calling the `result` function. (This callback design is chosen so that it's clear and easy to define memory ownership.)
+
+`assembly_path` is optional:
+* If not specified, the "muxer" (`dotnet.exe`) behavior is used to locate the `hostfxr`
+* If specified, the `apphost` behavior is used
+
+
+## Improve API to run application
+New API will be added to `hostfxr`.
+
+``` C++
+struct hostfxr_runtime_properties
+{
+    int count;
+    const char ** keys;
```

Comment:
There's prior use of `**Keys` and `**values` lists in core-setup code, but is it simpler to have key-value paired structs? 
Otherwise, `hostfxr_runtime_properties` and `hostfxr_parameters` could be merged into one struct.

Reply:
That sounds like a good idea.